### ISSUE TITLE
8.6 type inference behave different with 8.5

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#5379]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 5379
Date: 2017-03-03 03:58:38 +0100
From: Qinxiang Cao &lt;<caoqinxiang@gmail.com>&gt;
To:   
CC:   @_andrew-appel, coq-bugs-redist@lists.gforge.inria.fr, @_silene, @_JasonGross

See also:     https://coq.inria.fr/bugs/show_bug.cgi?id=4821
Last updated: 2017-03-07 09:13:34 +0100

## Bugzilla Comment ID: 15379
Date: 2017-03-03 03:58:38 +0100
From: Qinxiang Cao &lt;<caoqinxiang@gmail.com>&gt;

8.6 Type inference behave different with 8.5. Here is an example:

==========================================================
Fixpoint bool_tuple (l: nat): Type :=
  match l with
  | 0 => unit
  | S l' => bool * bool_tuple l'
  end.

Parameter P: forall l, bool_tuple (S l) -> Prop.

Goal forall l x, (fun t => P l (fst t, snd t)) x.
==========================================================

In Coq 8.5, the type of x is:
      bool * bool_tuple l.

In Coq 8.6, the type of x is:
      bool *
       (fix bool_tuple (l0 : nat) : Type :=
          match l0 with
          | 0 => unit
          | S l' => (bool * bool_tuple l')%type
          end) l)

Is this a bug or a feature of Coq 8.6?

## Bugzilla Comment ID: 15380
Date: 2017-03-03 08:42:35 +0100
From: @_silene

Note that, on your example, type inference computes the exact same type in both versions. What is somewhat broken is the refolding of recursive definitions to make types look nicer. You can perform it manually by typing "fold bool_tuple".

## Bugzilla Comment ID: 15381
Date: 2017-03-03 09:03:03 +0100
From: Qinxiang Cao &lt;<caoqinxiang@gmail.com>&gt;

(In reply to Guillaume Melquiond from comment #1)
> Note that, on your example, type inference computes the exact same type in
> both versions. What is somewhat broken is the refolding of recursive
> definitions to make types look nicer. You can perform it manually by typing
> "fold bool_tuple".

You are right. The inferred type must be equivalent w.r.t. beta-eta-delta-iota reduction, or else one of them will not type check.

I hope Coq can fix this nice refolding in type inference, or else I have to manually type a lot of definitions (which almost means type inference fails).

## Bugzilla Comment ID: 15382
Date: 2017-03-03 17:47:53 +0100
From: @_JasonGross

If I recall correctly, PMP (and/or Matthieu) found that refolding was an enormous performance bottleneck in [apply] in some cases in fiat-parsers.  See bug #4821.

## Bugzilla Comment ID: 15383
Date: 2017-03-03 18:00:39 +0100
From: Qinxiang Cao &lt;<caoqinxiang@gmail.com>&gt;

(In reply to Jason Gross from comment #3)
> If I recall correctly, PMP (and/or Matthieu) found that refolding was an
> enormous performance bottleneck in [apply] in some cases in fiat-parsers. 
> See bug #4821.

I think there is two problem here. One is the type inference in a definition or a lemma. The other is the ugly print out result in proof mode.

The second problem can be fixed by manually refolding. I think this is not too bad.

However, a different type inference result of a definition will cause some proof script of its property fails (which did work in Coq 8.5).

So, if a nicer refolding will cause performance problem in some specific tactics and cannot be fixed, I am OK with that. But I still hope that at least the type inference of definitions and lemmas can be fixed.

## Bugzilla Comment ID: 15389
Date: 2017-03-06 21:17:55 +0100
From: @_andrew-appel

Let me explain why I think this is a significant bug in Coq 8.6, and it is worth fixing.  

In the original example, suppose I have:

Lemma my_lemma:  forall l x, (fun t => P l (fst t, snd t)) x.
Proof.
 . . .
  (*  Oops:  the type of x is very messy,
     makes it hard to prove things.  This is true inside
   the proof . . . *)
Qed.

Lemma next_lemma: whatever.
Proof.
  (* . . . but it's also true here, when I try to _use_ 
   my_lemma in another proof.  Especially, the "auto" and "rewrite"
   tactics will not succeed when the types are unfolded like this. *)
Qed.

Lemma my_lemma':  forall l (x: bool * bool_tuple l),  (fun t => P l (fst t, snd t)) x.
Proof.
  (* Here, things are not messy, either inside the proof of my_lemma',
  or outside.  Great!  But we should not force the user to put in this
  type specification for x, especially since Coq 8.5 did this correctly!
 *)
Qed.

Summary:  Please try to refold definitions when occurring in parameters
  of top-level Definitions, Lemmas, etc.

## Bugzilla Comment ID: 15390
Date: 2017-03-06 21:56:11 +0100
From: @_JasonGross

Would this suggestion fix the problem?  When using a [Definition] or equivalent vernacular, add a post-processing phase that refolds fixpoints.

## Bugzilla Comment ID: 15391
Date: 2017-03-07 09:13:34 +0100
From: Qinxiang Cao &lt;<caoqinxiang@gmail.com>&gt;

(In reply to Jason Gross from comment #6)
> Would this suggestion fix the problem?  When using a [Definition] or
> equivalent vernacular, add a post-processing phase that refolds fixpoints.

I am not sure whether I understand "post-processing phase".

If this "post-processing phase" modify the type of the [Definition], then it works. If this "post-processing phase" only prints out nice proof goal, then it does not work.

In other words, if I do:

-----------------------------------------------------------------
Lemma my_lemma:  forall l x, (fun t => P l (fst t, snd t)) x.
Proof. ... Qed.

Check my_lemma.
-----------------------------------------------------------------

Then I hope that the type of x here is folded.
